docs/design/ha_master.md

```diff
@@ -0,0 +1,267 @@
+  1. Create load balancer using ephemeral IP address
+  2. Add existing apiserver to the load balancer
+  3. Wait until load balancer is working, i.e. all data is propagated, in GCE up to 20 min (sic!)
+  4. Update DNS to point to the load balancer.
+2. Clone existing master (create a new VM with the same configuration) including
+   all env variables (certificates, IP ranges etc), with the exception of
+   `INITIAL_ETCD_CLUSTER`.
+3. SSH to an existing master and run the following command to extend etcd cluster
+   with the new instance:
+   `curl <existing_master>:4001/v2/members -XPOST -H "Content-Type: application/json" -d '{"peerURLs":["http://<new_master>:2380"]}'`
+4. Add IP address of the new apiserver to the load balancer.
+```
+
+A simplified algorithm for adding a new master replica and promoting master IP to the load balancer
+is identical to the one when using DNS, with a different step to setup load balancer:
+
+```
+1. If there is no load balancer for this cluster:
+  1. Unassign IP from the existing master replica
+  2. Create load balancer using static IP reclaimed in the previous step
+  3. Add existing apiserver to the load balancer
+  4. Wait until load balancer is working, i.e. all data is propagated, in GCE up to 20 min (sic!)
+...
+```
+
+## Deleting replica
+
+Command to delete one replica on GCE using kube-up script:
+
+```
+KUBE_DELETE_NODES=false KUBE_GCE_ZONE=us-central1-b kubernetes/cluster/kube-down.sh
+```
+
+A simplified algorithm for deleting an existing replica for the master is the following:
+
+```
+1. Remove replica IP address from the load balancer or DNS configuration
+2. SSH to one of the remaining masters and run the following command to remove replica from the cluster:
+  `curl etcd-0:4001/v2/members/<id> -XDELETE -L`
+3. Delete replica VM
+4. If load balancer has only a single target instance, then delete load balancer
+5. Update DNS to point to the remaining master replica, or [on GCE] assign static IP back to the master VM.
+```
+
+## Upgrades
+
+Upgrading replicated master will be possible by upgrading them one by one using existing tools
+(e.g. upgrade.sh for GCE). This will work out of the box because:
+* Requests from nodes will be correctly served by either new or old master because apiserver is backward compatible.
+* Requests from scheduler (and controllers) go to a local apiserver via localhost interface, so both components
+will be in the same version.
+* Apiserver talks only to a local etcd replica which will be in a compatible version
+* Etcd replicas have backward compatible clustering & communication protocol so they will be able to share
+state even when they are in different versions, including upgrade from `v2` to `v3`. The only problem which remains is with
+upgrading storage format for etcd from v2 to v3, and hence upgrading APIs. This however is unrelated to
+HA master and will be discussed in a separate document, specific to etcd v3 upgrade.
```

Comment:
I'm not sure that @mml is going to worry about upgrading etcd2- > etcd3 in this sort of HA configuration (I think it's hard enough in a single master config), so I disagree that it's unrelated. 

One option, as you mentioned in comments above, is to take down all of the masters at once during an upgrade instead of leaving 2 up and taking one down. This will cause downtime (kind of the antithesis of HA) but may be necessary for some version upgrades.